sources/packages/backend/apps/api/src/services/application/application.service.ts

```diff
@@ -1517,6 +1521,142 @@ export class ApplicationService extends RecordDataModelService<Application> {
     });
   }
 
+  /**
+   * Gets the application details belonging to an institution location.
+   * @param locationId location id.
+   * @param applicationId application id.
+   * @returns student application.
+   */
+  async getApplicationInfo(
+    locationId: number,
+    applicationId: number,
+  ): Promise<Application> {
+    return this.repo.findOne({
+      select: {
+        id: true,
+        data: true as unknown,
+        programYear: {
+          id: true,
+          active: true,
+          startDate: true,
+          endDate: true,
+        },
+        student: {
+          id: true,
+          birthDate: true,
+          user: {
+            id: true,
+            lastName: true,
+          },
+          sinValidation: {
+            id: true,
+            sin: true,
+          },
+        },
+        currentAssessment: {
+          id: true,
+          offering: {
+            id: true,
+            offeringIntensity: true,
+          },
+        },
+      },
+      relations: {
+        programYear: true,
+        location: true,
+        student: {
+          user: true,
+          sinValidation: true,
+        },
+        currentAssessment: {
+          offering: true,
+        },
+      },
+      where: {
+        id: applicationId,
+        location: {
+          id: locationId,
+        },
+        applicationStatus: Not(ApplicationStatus.Overwritten),
+      },
+    });
+  }
+
+  /**
+   * Set of validations for existing application and newly
+   * selected offering, like, application belongs to the location,
+   * newly selected offering belongs to the location, is there any
+   * overlapping study dates for the student, is the offering is
+   * matching with the existing offering and the newly selected
+   * offering belong to the same program year.
+   * @param applicationId existing application id.
+   * @param selectedOffering newly selected offering id.
+   * @param locationId location id.
+   * @param options more options of the validation.
+   * - `isPir` set true, if its PIR related.
+   */
+  async applicationOfferingValidation(
+    applicationId: number,
+    selectedOffering: number,
+    locationId: number,
+    options?: {
+      isPir: boolean;
+    },
+  ): Promise<void> {
+    // Validate if the application exists and the location has access to it.
+    const application = options?.isPir
+      ? await this.getProgramInfoRequest(locationId, applicationId)
+      : await this.getApplicationInfo(locationId, applicationId);
+
+    if (!application) {
+      throw new CustomNamedError(
+        "Application not found.",
+        APPLICATION_NOT_FOUND,
+      );
+    }
+    // Validates if the offering exists and belongs to the location.
+    const offering = await this.offeringService.getOfferingLocationId(
+      selectedOffering,
+    );
+    if (offering?.institutionLocation.id !== locationId) {
```

Comment:
May be we should check if offering exist before doing this validation? Because if the offeringId is not a valid one `offering not found ` would be more accurate